File: ios/ICEInfo/Utilities/JourneyRepository.swift
import Foundation

actor JourneyRepository {
    static let shared = JourneyRepository()

    private let defaults = UserDefaults.standard
    private let journeysKey = "saved_journeys"

    func loadJourneys() -> [SavedJourney] {
        guard let data = defaults.data(forKey: journeysKey) else { return sampleJourneys }
        let journeys = (try? JSONDecoder().decode([SavedJourney].self, from: data)) ?? []
        return journeys.isEmpty ? sampleJourneys : journeys
    }

    func saveJourney(_ journey: SavedJourney) {
        var current = loadJourneys()
        current.insert(journey, at: 0)
        if let data = try? JSONEncoder().encode(current) {
            defaults.set(data, forKey: journeysKey)
        }
    }

    func deleteJourney(id: String) {
        var current = loadJourneys()
        current.removeAll { $0.id == id }
        if let data = try? JSONEncoder().encode(current) {
            defaults.set(data, forKey: journeysKey)
        }
    }
}

File: app/src/main/java/com/nruge/iceinfo/JourneyRepository.kt
package com.nruge.iceinfo

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.nruge.iceinfo.model.SavedJourney
import kotlinx.coroutines.flow.first
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

private val Context.journeyDataStore by preferencesDataStore(name = "journeys")

object JourneyRepository {

    private val JOURNEYS_KEY = stringPreferencesKey("saved_journeys")
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun loadJourneys(context: Context): List<SavedJourney> {
        val prefs = context.journeyDataStore.data.first()
        val raw = prefs[JOURNEYS_KEY] ?: return emptyList()
        return runCatching { json.decodeFromString<List<SavedJourney>>(raw) }
            .getOrDefault(emptyList())
    }

    suspend fun saveJourney(context: Context, journey: SavedJourney) {
        context.journeyDataStore.edit { prefs ->
            val current = runCatching {
                json.decodeFromString<List<SavedJourney>>(prefs[JOURNEYS_KEY] ?: "[]")
            }.getOrDefault(emptyList())
            prefs[JOURNEYS_KEY] = json.encodeToString(listOf(journey) + current)
        }
    }

    suspend fun deleteJourney(context: Context, id: String) {
        context.journeyDataStore.edit { prefs ->
            val current = runCatching {
                json.decodeFromString<List<SavedJourney>>(prefs[JOURNEYS_KEY] ?: "[]")
            }.getOrDefault(emptyList())
            prefs[JOURNEYS_KEY] = json.encodeToString(current.filter { it.id != id })
        }
    }
}
